drivers: gpio: shell: improve usability

nixward opened the pull request:
This commit implements this enhancement:
https://github.com/zephyrproject-rtos/zephyr/issues/63018

Also adds a 'info' command for ngpios, reversed pin and line
name information.

The forms of the gpio commands are now:

        gpio conf device pin ol0
        gpio set device pin 1
        gpio get device pin
        gpio blink device pin
        gpio info device

Device name, pin, configuration and set level sub commands now
are suggested/completed when tab is used.

Pin names are suggested with numbers and if available from the Devicetree
gpio line names.

GPIO pin control is now limited to pins that are not assigned as reserved.

Fixes #63018

nixward commented:
@fabiobaltieri would you mind having a look at this?

nixward commented:
@nandojve would you mind having a look as well?

nixward commented:
@aasinclair I added a commit to add more gpio configuration options, I'm not sure it's exhaustive but it's expanded (and maybe the most useful). See what you think.

aasinclair commented:
Hi @nixward.
Would it be possible to support a custom value, maybe something like:
`uart:~$ gpio gpio@50000000 17 rawconf 0x10000`

Implemented like this?
`static int cmd_rawconf(const struct shell *sh, size_t argc, char **argv, void *data)
{
	struct sh_gpio gpio;
	uint32_t flags;
	int ret = 0;

	get_sh_gpio(sh, argv, -2, &gpio);

	flags = shell_strtoul(argv[1], 0, &ret);
	if (ret != 0) {
		shell_error(sh, "error: %d", ret);
		return ret;
	}

	ret = gpio_pin_configure(gpio.dev, gpio.pin, flags);
	if (ret != 0) {
		shell_error(sh, "error: %d", ret);
		return ret;
	}

	return 0;
}`

nixward commented:
Hi @aasinclair I've been reading the gpio API headers `include/zephyr/drivers/gpio.h` and `include/zephyr/dt-bindings/gpio.h` should only flags from the set `(GPIO_INPUT | GPIO_OUTPUT | GPIO_DT_FLAGS_MASK)` be allowed? To me it doesn't make sense to allow flags from `GPIO_INT_MASK`.

aasinclair commented:
> Hi @aasinclair I've been reading the gpio API headers `include/zephyr/drivers/gpio.h` and `include/zephyr/dt-bindings/gpio.h` should only flags from the set `(GPIO_INPUT | GPIO_OUTPUT | GPIO_DT_FLAGS_MASK)` be allowed? To me it doesn't make sense to allow flags from `GPIO_INT_MASK`.

Hi @nixward, allowing a custom value is most useful for device specific configuration (driver strength, debounce etc).
E.g.:
https://github.com/zephyrproject-rtos/zephyr/blob/main/include/zephyr/dt-bindings/gpio/nuvoton-npcx-gpio.h
or
https://github.com/zephyrproject-rtos/zephyr/blob/main/include/zephyr/dt-bindings/gpio/atmel-sam-gpio.h

keith-zephyr commented on drivers/gpio/gpio_shell.c:
```diff
@@ -21,167 +21,161 @@
-		shell_error(sh, "Wrong parameters for get");
-		return -EINVAL;
-	}
+	get_sh_gpio(sh, argv, -2, &gpio);
```
Please add some constants so you can calculate the argv index.
For instance:
```
#define ARGV_GPIO_PORT 1
#define ARGV_GPIO_PIN 2
#define ARGV_SUBCMD 3
#define ARGV_SET_VALUE 4
```

Then this call here becomes
```
	get_sh_gpio(sh, argv, ARGV_GPIO_PORT - ARGV_GPIO_SUBCMD, &gpio);
```

nixward replied:
Nice one, I'll add 👍

marco85mars20 replied:
By reshuffling the args, it would be possible to remove the table `pin_syntax` and make it as a standard manual entry:
```
#define ARGV_SUBCMD 1
#define ARGV_SET_VALUE 2
#define ARGV_GPIO_PORT 3
#define ARGV_GPIO_PIN 4

keith-zephyr commented on drivers/gpio/gpio_shell.c:
```diff
@@ -21,167 +21,161 @@
-			size_t argc, char **argv)
+SHELL_SUBCMD_DICT_SET_CREATE(sub_conf, conf_cmd_handler,
+	(in, GPIO_INPUT, "input"),
+	(inu, (GPIO_INPUT | GPIO_PULL_UP), "input with pull up"),
+	(ind, (GPIO_INPUT | GPIO_PULL_DOWN), "input with pull down"),
```
Non-blocking suggestion:
```suggestion
	(in_pu, (GPIO_INPUT | GPIO_PULL_UP), "input with pull up"),
	(in_pd, (GPIO_INPUT | GPIO_PULL_DOWN), "input with pull down"),
```

keith-zephyr commented on drivers/gpio/gpio_shell.c:
```diff
@@ -53,11 +53,18 @@ static int conf_cmd_handler(const struct shell *sh,
 	(ind, (GPIO_INPUT | GPIO_PULL_DOWN), "input with pull down"),
-	(out, GPIO_OUTPUT, "output")
+	(out, GPIO_OUTPUT, "output"),
+	(al_in, (GPIO_ACTIVE_LOW_INPUT), "active low input"),
+	(al_inu, (GPIO_ACTIVE_LOW_INPUT | GPIO_PULL_UP), "active low input with pull up"),
+	(al_ind, (GPIO_ACTIVE_LOW_INPUT | GPIO_PULL_DOWN), "active low input with pull down"),
+	(al_out, (GPIO_ACTIVE_LOW_OUTPUT), "active low output")
```
Two alternatives:
```suggestion
	(al_in, (GPIO_ACTIVE_LOW_INPUT), "active low input"),
	(al_in_pu, (GPIO_ACTIVE_LOW_INPUT | GPIO_PULL_UP), "active low input with pull up"),
	(al_in_pd, (GPIO_ACTIVE_LOW_INPUT | GPIO_PULL_DOWN), "active low input with pull down"),
	(al_out, (GPIO_ACTIVE_LOW_OUTPUT), "active low output")
```
```suggestion
	(in#, (GPIO_ACTIVE_LOW_INPUT), "active low input"),
	(in_pu#, (GPIO_ACTIVE_LOW_INPUT | GPIO_PULL_UP), "active low input with pull up"),
	(in_pd#, (GPIO_ACTIVE_LOW_INPUT | GPIO_PULL_DOWN), "active low input with pull down"),
	(out#, (GPIO_ACTIVE_LOW_OUTPUT), "active low output")
```

marco85mars20 commented on drivers/gpio/gpio_shell.c:
```diff
@@ -21,167 +21,168 @@
 			       );
 
-SHELL_CMD_REGISTER(gpio, &sub_gpio, "GPIO commands", NULL);
+static const char * const pin_syntax[] = {
+	 "0",  "1",  "2",  "3",  "4",  "5",  "6",  "7",  "8",  "9",
+	"10", "11", "12", "13", "14", "15", "16", "17", "18", "19",
+	"20", "21", "22", "23", "24", "25", "26", "27", "28", "29",
+	"30", "31"
+};
+
+static void cmd_gpio_pin_get(size_t idx, struct shell_static_entry *entry)
+{
+	if (idx < ARRAY_SIZE(pin_syntax)) {
+		entry->syntax  = pin_syntax[idx];
+		entry->handler = NULL;
+		entry->subcmd = &sub_gpio;
+		entry->help = "Pin";
+	} else {
+		entry->syntax = NULL;
+		entry->subcmd = &sub_gpio;
+	}
+}
```
Would it be possible to remove `pin_syntax` and generate the list of available pins on the fly based on `DT_PROP(DT_NODELABEL(dev->name), ngpios)` ?

```suggestion
static void cmd_gpio_pin_get(size_t idx, struct shell_static_entry *entry)
{
	struct sh_gpio gpio;
	uint8_t ngpios;

	/* NOTE: Not sure how to retrieve 'sh' */

	get_sh_gpio(sh, argv, ARGV_GPIO_PORT - ARGV_GPIO_PIN, &gpio);
	ngpios = DT_PROP_OR(DT_NODELABEL(gpio.dev->name), ngpios, 0);

	if (idx < ngpios) {
		entry->syntax  = pin_syntax[idx];
		entry->handler = NULL;
		entry->subcmd = &sub_gpio;
		entry->help = "Pin";
	} else {
		entry->syntax = NULL;
		entry->subcmd = &sub_gpio;
	}
}
```

nixward replied:
@marco85mars20 yes, similar to this I've been using the Devicetree to generate the pin argument code, I'll hopefully push soon.

marco85mars20 reviewed (commented):
@nixward what do you think about these suggestions?

nixward committed:
boards: nrf52840dk_nrf52840: add gpio reserved ranges and line names

Add gpio-reserved-ranges and gpio-line-names properties.

Signed-off-by: Nick Ward <nix.ward@gmail.com>

nixward committed:
drivers: gpio: shell: improve tab complete/suggestion support

This commit implements this enhancement:
https://github.com/zephyrproject-rtos/zephyr/issues/63018

The forms of the gpio commands are now:

        gpio conf device pin ol0
        gpio set device pin 1
        gpio get device pin
        gpio blink device pin

Device name and pin subcommands now are
suggested/completed when tab is used.

Pin names are suggested with numbers and line names if
available from the gpio controller’s Devicetree node.

GPIO pin command is now limited to pins that are not assigned
as reserved.

Signed-off-by: Nick Ward <nix.ward@gmail.com>

nixward commented:
Tab complete with line name for gpio conf and get commands:
![image](https://github.com/zephyrproject-rtos/zephyr/assets/131832/04fc6919-2c7f-42aa-ab4a-7ade167b48ff)

Info command providing device:
![image](https://github.com/zephyrproject-rtos/zephyr/assets/131832/40a74d50-075d-4ead-ae48-c25e7f97ffe8)

Info command without providing device:
![image](https://github.com/zephyrproject-rtos/zephyr/assets/131832/55699188-3f35-48f3-9f7c-a97cb3b9d628)

nixward commented:
I've added a new sample for Arduino useful shell commands. Happy to remove this is this is overkill and instead add gpio and adc shell (adc shell doesn't seem to be currently built in CI) command access to another sample with shell enabled.

nixward committed:
drivers: gpio: shell: add info command

Usage:
gpio info [device]

The new command prints gpio controller information
for a specific device if specified or if no device is specified
it prints out all controller information ordered by line name.

Also added Kconfig option so this command can be removed if
resources need to be conserved.

Signed-off-by: Nick Ward <nix.ward@gmail.com>

nixward committed:
drivers: gpio: shell: make blink command optional

Adds CONFIG_GPIO_SHELL_BLINK_CMD symbol.
Saves around 300 bytes when command is disabled.

Signed-off-by: Nick Ward <nix.ward@gmail.com>

nixward committed:
drivers: gpio: shell: optimise blink command

Use gpio toggle api instead of manually toggling.
Remove redundant text.
Print error and break from blinking if it occurs.
Only print 'how to exit' text if first toggle is successful.

Saves roughly 40 bytes.

Signed-off-by: Nick Ward <nix.ward@gmail.com>

nixward committed:
drivers: gpio: shell: add vendor specific flags argument

Allow the optional setting of vendor specific flags in the conf command.

Signed-off-by: Nick Ward <nix.ward@gmail.com>

nixward commented:
Hi @aasinclair, after reading in zephyr/include/zephyr/dt-bindings/gpio/ and it's [history](https://github.com/zephyrproject-rtos/zephyr/commit/2533b13cd21ddc08bec5fd1cc0018ec470eea596) I now understand the vendor specific flags are in the mask 0xFF00 so I noticed we could extend the current conf command to add this feature.  See what you think in commit: https://github.com/zephyrproject-rtos/zephyr/pull/63019/commits/7d8277bfb2fb30dcfb5188b1f78caf2349d7b64e

aasinclair commented on drivers/gpio/gpio_shell.c:
```diff
@@ -1,187 +1,614 @@
+
+	if (argc == 5) {
+		vendor_specific = shell_strtol(argv[ARGV_VENDOR_SPECIFIC], 0, &ret);
+		if ((vendor_specific & ~(0xFF00U)) == 0) {
```
```suggestion
		if ((ret == 0) && ((vendor_specific & ~(0xFF00U)) == 0)) {
```

aasinclair reviewed (commented):
The optional vendor config argument is great.
One odd thing I see is a two strange options at the end of the pin list:

```
uart:~$ gpio get gpio@50000000 
  0   1   2   3   4   5   6   7   8   9   10  11  12  13  14  15  16  17  18
  19  20  21  22  23  24  25  26  27  28  29  30  31  �D  �D
```

nixward commented:
> The optional vendor config argument is great. One odd thing I see is a two strange options at the end of the pin list:
> 
> ```
> uart:~$ gpio get gpio@50000000 
>   0   1   2   3   4   5   6   7   8   9   10  11  12  13  14  15  16  17  18
>   19  20  21  22  23  24  25  26  27  28  29  30  31  �D  �D
> ```

Thanks @aasinclair this bug is now fixed.

keith-zephyr commented on drivers/gpio/gpio_shell.c:
```diff
@@ -425,6 +424,147 @@ static void device_name_get(size_t idx, struct shell_static_entry *entry)
+	size_t i;
+
+	for (i = 0; i < ARRAY_SIZE(gpio_list); i++) {
+		for (pin = 0; pin < GPIO_MAX_PINS_PER_PORT; pin++) {
```
```suggestion
		for (pin = 0; pin < gpio_list[i].ngpios; pin++) {
```

nixward replied:
Done

nixward commented on drivers/gpio/Kconfig:
```diff
@@ -16,6 +16,7 @@ source "subsys/logging/Kconfig.template.log_config"
 
 config GPIO_SHELL
 	bool "GPIO Shell"
+	default y
```
Hi Keith, thanks for the review. Sorry for my confusion, are you referring to the PR description above (I've now added) or the next Zephyr release notes?

cfriedt replied:
Is defaulting it to `y` necessary?

aasinclair replied:
I believe I asked for this originally.
With this change enabling SHELL will automatically enable GPIO_SHELL, rather than requiring both to specified.

This seems to be the default for newer shell interfaces, but there are several older ones which require it explicitly.

nixward replied:
@keith-zephyr I've updated the v3.6 release notes.

aasinclair replied:
There has been further discussion about default behaviour of shells on discord leading to this PR:
https://github.com/zephyrproject-rtos/zephyr/pull/65307

So I suggest removing this change (sorry).

nixward commented on boards/arm/nrf52840dk_nrf52840/nrf52840dk_nrf52840.dts:
```diff
@@ -149,10 +149,18 @@
 &gpio0 {
 	status = "okay";
+	gpio-reserved-ranges = <0 11>, <17 7>, <26 6>;
+	gpio-line-names = "XL1", "XL2", "AREF", "A0", "A1", "UART RTS", "UART TxD", "UART CTS",
+	"UART RxD", "NFC1", "NFC2", "BUTTON1", "BUTTON2", "LED1", "LED2", "LED3",
+	"LED4", "QSPI CS", "RESET", "QSPI CLK", "QSPI0", "QSPI1", "QSPI2", "QSPI3",
+	"BUTTON3", "BUTTON4", "I2C SDA", "I2C SCL", "A2", "A3", "A4", "A5";
```
Good question, I was following the Linux Devicetree guide https://www.kernel.org/doc/Documentation/devicetree/bindings/gpio/gpio.txt and in it they say:

> In either case placeholders are discouraged: rather use the "" (blank
string) if the use of the GPIO line is undefined in your design. Ideally,
try to add comments to the dts file describing the naming the convention
you have chosen, and specifying from where the names are derived.

I missed the detail in that. I'll replace the non GPIO pin line names so there's only LED1-4, BUTTON1-4 and the P1.09 remaining. The new info command will be less useful but this fix will match the binding guide.

nixward replied:
Done

nixward commented on drivers/gpio/gpio_shell.c:
```diff
@@ -1,187 +1,449 @@
+ * Find idx-th pin reference from the set of non reserved
+ * pin numbers and provided line names.
+ */
+static void port_pin_get(gpio_port_pins_t reserved_mask, const char **line_names,
+			 uint8_t line_names_len, size_t idx, struct shell_static_entry *entry)
```
Yes true, I was thinking at the time it would be good to give the user the option pin number or line name but at the same time during testing I stopped using the pin numbers and was using line names exclusively. Happy to use line name exclusively when available and fall back to pin numbers when line name is not available unless anyone else disagrees.

nixward commented on drivers/gpio/gpio_shell.c:
```diff
@@ -1,187 +1,449 @@
+		}
+		pin_syntax[sizeof(pin_syntax) - 1] = '\0';
+		entry->syntax = pin_syntax;
+		entry->help = line_names[pin];
```
Good idea, can do 👍

nixward commented on drivers/gpio/gpio_shell.c:
```diff
@@ -1,187 +1,449 @@
+	int pin;
+
+	gpio->dev = device_get_binding(argv[ARGV_DEV]);
+	ctrl = get_gpio_ctrl(argv[ARGV_DEV]);
```
Done

nixward commented on drivers/gpio/gpio_shell.c:
```diff
@@ -1,187 +1,449 @@
+		shell_error(sh, "unknown gpio controller: %s", argv[ARGV_DEV]);
 		return -EINVAL;
 	}
+	pin = shell_strtoul(argv[ARGV_PIN], 0, &ret);
```
Fixed

keith-zephyr commented on drivers/gpio/gpio_shell.c:
```diff
@@ -221,6 +222,7 @@ static int get_sh_gpio(const struct shell *sh, char **argv, struct sh_gpio *gpio
 static int cmd_gpio_conf(const struct shell *sh, size_t argc, char **argv, void *data)
 {
 	gpio_flags_t flags = 0;
+	long vendor_specific;
```
Why are you using `long` instead of `gpio_flags_t` here?

nixward replied:
Fixed

keith-zephyr reviewed (commented):
Please also fix the compliance checks.

nixward commented on drivers/gpio/gpio_shell.c:
```diff
@@ -1,187 +1,441 @@
+			/* Allow underscore to match space */
+		} else if (input[i] != line_name[i]) {
+			return (input[i] > line_name[i]) ? 1 : -1;
+		} else if (line_name[i] == '\0') {
```
This check happens in the logic. If the strings are different at any stage, i.e. `input[i]=='\0'` and `line_name[i]` doesn't it will exit otherwise if the characters are equal and `line_name[i]=='\0'` it is a match.

nixward commented on boards/arm/nrf52840dk_nrf52840/nrf52840dk_nrf52840.dts:
```diff
@@ -149,10 +149,18 @@
+	gpio-line-names = "", "", "", "", "", "", "", "",
+	"", "", "", "BUTTON1", "BUTTON2", "LED1", "LED2", "LED3",
+	"LED4", "", "", "", "", "", "", "",
+	"BUTTON3", "BUTTON4", "", "", "", "", "", "";
```
Done

nixward commented on boards/arm/nrf52840dk_nrf52840/nrf52840dk_nrf52840.dts:
```diff
@@ -149,10 +149,18 @@
 	status = "okay";
+	gpio-reserved-ranges = <0 1>, <12 4>;
+	gpio-line-names = "", "D0", "D1", "D2", "D3", "D4", "D5", "D6",
+	"D7", "P1.09", "D8", "D9", "", "", "", "";
```
Done

keith-zephyr commented on boards/arm/nrf52840dk_nrf52840/nrf52840dk_nrf52840.dts:
```diff
@@ -149,10 +149,18 @@
 
 &gpio1 {
 	status = "okay";
+	gpio-reserved-ranges = <0 1>, <12 4>;
+	gpio-line-names = "", "D0", "D1", "D2", "D3", "D4", "D5", "D6",
+		"D7", "", "D8", "D9", "", "", "", "";
```
Do you need a reserved range for `<9 1>`?

nixward replied:
Fixed

keith-zephyr commented on drivers/gpio/gpio_shell.c:
```diff
@@ -1,187 +1,441 @@
-		} else {
-			shell_error(sh, "Error %d reading value", rc);
-			return -EIO;
+	if (((flags & GPIO_INPUT) != 0) == ((flags & GPIO_OUTPUT) != 0)) {
+		shell_error(sh, "must be either input or output");
+		shell_help(sh);
+		return SHELL_CMD_HELP_PRINTED;
+	}
```
Non-blocking comment - This duplicates the logic from `gpio_pin_configure()`, which uses asserts.
As a follow up CL, you could create a helper function for this validation that is used both by the shell command and `gpio_pin_configure()`.

keith-zephyr commented on drivers/adc/adc_ads114s0x.c:
```diff
@@ -1080,7 +1080,7 @@ static int ads114s0x_gpio_write_config(const struct device *dev)
 	struct ads114s0x_data *data = dev->data;
 	const struct ads114s0x_config *config = dev->config;
-	uint8_t register_addresses[2];
+	enum ads114s0x_register register_addresses[2];
```
This commit and the following commits are not related to the GPIO shell changes.  Please move to a separate PR.

nixward replied:
This issue was was one of many with the ads114s0x_gpio driver/test that were causing CI to fail for this PR. I've opened https://github.com/zephyrproject-rtos/zephyr/pull/64876

keith-zephyr replied:
https://github.com/zephyrproject-rtos/zephyr/pull/64876 has merged.  Please rebase this PR.

nixward replied:
Done 👍

keith-zephyr reviewed (commented):
Looks good to me once you drop the top two commits that enable the GPIO shell by default.

nixward commented:
> Looks good to me once you drop the top two commits that enable the GPIO shell by default.

Done
